MdeModulePkg/Universal/FaultTolerantWriteDxe/FtwMisc.c

```diff
@@ -815,7 +815,10 @@ FtwGetLastWriteHeader (
     //
     // If Offset exceed the FTW work space boudary, return error.
     //
-    if (Offset >= FtwWorkSpaceSize) {
+     // MU_CHANGE [BEGIN] - Prevent buffer overrun
+    //if (Offset >= FtwWorkSpaceSize) {
+    if ((Offset + sizeof (EFI_FAULT_TOLERANT_WRITE_HEADER)) >= FtwWorkSpaceSize)

```

Comment:
Is it really illegitimate for the "=" condition? Can it not just be an empty header?

Reply:
Without including the `EFI_FAULT_TOLERANT_WRITE_HEADER` the condition used to be `>=`, was this inaccurate?

Reply:
@sureshkumarpMSFT, I'm good to go with Kun's suggestion. @VivianNK, yeah, the assumption here is that this was incorrect.

Reply:
This would reintroduce a bug fixed 12 years ago - https://github.com/tianocore/edk2/commit/fe92f438ac860a09e57a2a6a37c1a1d8b6539321.

Reply:
Thank you @makubacki for finding this history. Much appreciated

Reply:
I understand this is changing the calculation, but this logic has been around long enough that we need a stronger assertion to make a change. @sureshkumarpMSFT, please double check this when you look at the flow in more detail for the other conversation thread.

Reply:
Hi @makubacki @kuqin12 @VivianNK @os-d 

I would like to take a step back in making the change to remove `=` condition.
since this logic is there for long time and served well. There was bug reported with having `=` condition removed.

For testing, tried to do

_*FtwWriteHeader = (EFI_FAULT_TOLERANT_WRITE_HEADER*)((UINTN) (FtwWorkSpaceHeader) + FtwWorkSpaceSize - sizeof (EFI_FAULT_TOLERANT_WRITE_HEADER));_

and I do not see any hang or exception.

 But still, I am leaning towards not removing the `=` condition. We do not know exact scenario for the change https://github.com/tianocore/edk2/commit/fe92f438ac860a09e57a2a6a37c1a1d8b6539321 .  So, it's better to have that code as it is.

If we find a compelling case and if needed, we can work it as a separate issue later.  Please let me know your thoughts.

Reply:
I agree. Any changes need to have a detailed assessment of the problematic scenario (including diagrams with correlation to the overall flow) and subsequent cause and effect described by the proposed change.

This code can cause silent bugs to not be caught until late in the field.

Really, a unit test demonstrating the problem and showing the fix would be ideal but a heavy lift given no existing FTW unit test code.

Reply:
Agreed, thanks for the due diligence.